### ISSUE TITLE
Question: Video Inference

### ISSUE THREAD

**Issue by psandev** [NVIDIA/DALI#3359]
Hi,
I bumped on an issue and I would like  your comment, please. 
I trained an action recognition model on 16-frames clips. The inference I performed on an uncut big videos and the results came out rubbish.
Then I cut the big video into 16-frames clips using OpenCV,  exactly as the dataloader cuts the batches and the results were perfect.
I am guessing this is related to the video compression, but this means that, the way I trained the model prevents me from using DALI for inference, which is very regrettable.
I do not want to start from scratch training on full videos, the annotation transformation of the dataset will take forever. 
Is there anything else you could suggest I do? Thank you very much.

**Comment by JanuszL** [comment]
Hi @psandev,

I'm sorry but this is not our area of expertise. Maybe you can ask on https://github.com/NVIDIA/DeepLearningExamples where our Deep Learning experts develop models.

Could you tell how exactly you did the inference with a big, uncut video? Do you still extract 16 frames at a time and feed them into the network?

From DALI's point of view, it doesn't make much difference (if any) if you have a lot of small files or one big. It still builds a list of all possible sequences, shuffles, and feeds into the network.

**Comment by awolant** [comment]
Hello @psandev, thanks for the question.
What you described is consistent with my experience working with FFMPEG. Basically, when you cut the video, even though frames are visibly the same, the values of particular pixels are different, sometimes quite significantly. You can replicate this with FFMPEG. This command:
```
ffmpeg -i test.mp4 %03d.png
```
splits the video `test.mp4` into separate frames. Then you can compare frames from the small videos and from the large one outside of DALI (using OpenCV or whatever tool). **Important disclaimer**: DALI uses NVDEC rather than FFMPEG to decode the video, so frames returned by DALI will again be slightly different than what you get from FFMPEG. Unfortunately, this is the nature of working with video at the moment.
To sum up, in this case, when you move from cut videos to uncut ones, you slightly shift the distribution of your input data. How this affect your inference results will depend from case to case.
Additionally, with the parameters you shared, when you do the inference on the uncut video, you present additional sequences to the model. The sequences that cross the boundary at which the big video was cut into the smaller ones. This also might affect the results you get.
In both cases, things to work around this issue is either train the model you have on the long videos (transfer-learning style) or move the inference data to the distribution of your training data (cut the video before the inference). 
Hope that helps a bit.

**Comment by awolant** [comment]
I suggest to start with the same tool, that was used to prepare training dataset. This way, you can avoid the issue of shifting distribution and confirm the end-to-end working of your system. 
Unfortunately, I don't have enough knowledge to make an informed suggestion with regards to performance of various tools. I used FFMPEG and as far as I know, most of other libraries are basically wrappers around it, so probably in makes sense to try it out and see what happens.

**Comment by psandev** [comment]
This is my first video project and looks like I have constrained myself from the beginning due to lack of experience with video. Thanks again for helping sorting this out. It took me quite a while to find the root cause.

**Comment by awolant** [comment]
No problem, happy to help. If you have any more questions around DALI in the future, feel free to post another issue.
